ui-modules/blueprint-composer/app/components/catalog-selector/catalog-selector.template.html

```diff
@@ -30,44 +30,83 @@
+              
+              <div class="spacer"></div>
+                
+              <span uib-dropdown>
+                <a href id="palette-sort" uib-dropdown-toggle aria-haspopup="true" aria-expanded="false" >
+                  <div class="nav-row-item tool" title="Sort">
+                    <i class="fa fa-sort"></i></div>
+                </a>
+                <ul class="dropdown-menu right-align-icon" role="menu" uib-dropdown-menu aria-labelledby="palette-sort">
+                        <li role="menuitem" ng-repeat="order in viewOrders track by $index" ng-class="{'active': state.currentOrder[0] === order.field}" class="layer">
+                            <a ng-click="sortBy(order)"><i class="fa fa-fw fa-circle"></i> {{ order.label }}</a>
+                        </li>
+                </ul>
+              </span>
+              
+              <span uib-dropdown>
+                <a href id="palette-view-mode" uib-dropdown-toggle aria-haspopup="true" aria-expanded="false" >
+                  <div class="nav-row-item tool" title="Display mode">
+                    <i class="fa fa-th"></i></div>
+                </a>
+                <ul class="dropdown-menu right-align-icon" role="menu" uib-dropdown-menu aria-labelledby="palette-view-mode">
+                        <li role="menuitem" ng-repeat="view in viewModes track by $index" ng-class="{'active': state.viewMode === view}" class="layer">
+                            <a ng-click="state.viewMode = view"><i class="fa fa-fw fa-circle"></i> {{view.name}}</a>
+                        </li>
+                </ul>
+              </span>
+            </div>
             <ng-include src="customSubHeadTemplateName"/>
         </div>
-        <small class="help-block text-sm no-match" ng-if="search && searchedItems.length === 0">
-            No {{family.displayName.toLowerCase()}} matching the current search
-        </small>
+      </div>
 
-        <div class="row grid catalog-palette">
+        <div class="row grid">
             <!-- here and below, col-xs-3 or -4 or -2 all work giving different densities;
                  this could be configurable ("compressed"=xs-2 w no labels, "normal"=xs-3, "big"=xs-4) -->
-            <div class="col-xs-3 catalog-palette-item"
-                    ng-repeat="item in searchedItems = (filterPaletteItems(items | catalogSelectorSearch:search) | catalogSelectorSort:family) | orderBy:state.currentOrder | limitTo:pagination.itemsPerPage:(pagination.page-1)*pagination.itemsPerPage track by (item.containingBundle + ':' + item.symbolicName + ':' + item.version)"
+            <div class="catalog-palette-item" ng-class="state.viewMode.classes"
+                    ng-repeat="item in searchedItems = filterPaletteItemsWithActiveFilters(items | catalogSelectorSearch:search) | orderBy:state.currentOrder | limitTo:pagination.itemsPerPage:(pagination.page-1)*pagination.itemsPerPage track by (item.containingBundle + ':' + item.symbolicName + ':' + item.version)"
```

Comment:
It's a very bad angular practice to call for a function that returns a new array on every call: the track cannot run properly as it will be new items every time. Considering the number of items in the list, this should use an angular filter instead.

Reply:
noted - will work on this